Add timezone toggle to schedule calendar

bl4nkfce opened the pull request:
## Describe your changes

Resolves https://github.com/alveusgg/alveusgg/issues/592

- Updated timezone toggle styles and semantics
- Timezone is now stored in the localStorage

I've switched the checkbox into a button. This way we don't have to write custom logic for handling focus state in order to display an outline when navigating via keyboard. The HTML for this element is based on the [W3C APG example](https://www.w3.org/WAI/ARIA/apg/patterns/switch/examples/switch-button/) implementation.


### Desktop view
<img width="784" alt="image" src="https://github.com/user-attachments/assets/6cfefb8b-4cb4-4e2e-bf8d-b33ffb28c4aa">


It is a little cluttered on mobile, so maybe we can place the toggle between the arrows?

### Mobile (current)
<img width="350" alt="image" src="https://github.com/user-attachments/assets/09716eb3-f489-43a9-81ea-832ddd9084bc">

### Mobile (moved down a bit)
<img width="364" alt="image" src="https://github.com/user-attachments/assets/edda667d-7525-4a51-b8b0-95ae47ca7738">


## Notes for testing your change

1. Toggle the switch, check if there are any changes in the calendar.
2. Refresh the page and check if the toggle is in the same state as before.

Optionally check localStorage before and after refreshing the page to see if the timezone is stored correctly.

pjeweb commented:
Im a GH noob. Will this actually merge on my branch on not main? Can you make it a PR to the main branch instead so we can skip the double PR?

pjeweb commented:
@bl4nkfce this looks great. Is there an accessible way to lose the focus style after click? i always have to check again. Its just a bit confusing when the toggle switch stays dark after disabling it.

bl4nkfce commented:
> Im a GH noob. Will this actually merge on my branch and not main? Can you make it a PR to the main branch instead so we can skip the double PR?

Yeah, it would merge my branch into yours. I'll change the target to main.

---


> @bl4nkfce this looks great. Is there an accessible way to lose the focus style after click? i always have to check again. Its just a bit confusing when the toggle switch stays dark after disabling it.

Since we have the native focus outline applied to the element when navigating via keyboard, we don't need the additional dark background styles (we can remove the focus-within styles).

For desktop we can apply the "hover" styles only when the switch is not focused,
it would look like this:

https://github.com/user-attachments/assets/b532164b-6eb8-4a9a-860a-8853c17e3d62

pjeweb commented:
Since some part of this PR is from my brain, I'd love you, @MattIPv4, to review this when you have the time. I'll go through it too tho.

MattIPv4 commented:
Love this! One thing though, can we see if we can improve the clarity of the UX around changing between local/Alveus time? I'm not sure a toggle at the top is particularly clear what it is doing, nor that by default the schedule is local time?

Perhaps we could just have a line of text? `Using your local time (UTC), [switch to Alveus time (CT)](js click handler)` if that makes sense? (and flip the sentence when clicked)

MattIPv4 commented:
Oooh yeh, segmented button works too 👍 And feel free to play around with the locale stuff to have it work nicer

pjeweb reviewed (dismissed):
This looks great to me. Thank you

MattIPv4 commented on apps/website/src/components/calendar/Calendar.tsx:
```diff
@@ -138,6 +143,36 @@ const getCalendarTheme = (month?: number): CalendarTheme => {
+}
+
+export function useCalendarEventsQuery(selected?: MonthSelection) {
+  const start = useMemo(
+    () => selected && new Date(selected.year, selected.month, 1),
+    [selected],
+  );
+  const end = useMemo(
+    () => selected && new Date(selected.year, selected.month + 1, 1),
+    [selected],
+  );
```
Do we need to consider timezone here? If an event was right at the end of the day Alveus time on the last day of the month, it wouldn't show in this query based on local time, I think?

MattIPv4 replied:
![image](https://github.com/user-attachments/assets/fb69c107-78df-4c59-88db-7870db31a453)
![image](https://github.com/user-attachments/assets/1aaa03ca-0e27-4feb-9cb1-f095f9968c47)
![image](https://github.com/user-attachments/assets/c1e29b1e-097b-4eca-b8e1-2ec9c128f389)
![image](https://github.com/user-attachments/assets/77486da8-c0bb-4518-8fc4-115b6cf4b616)

MattIPv4 replied:
(As an aside, the fading that is applied to previous days doesn't seem to be timezone-aware)

MattIPv4 commented:
👋 @bl4nkfce any update on this?

bl4nkfce commented:
I looked at this briefly today.
I noticed we're using some luxon utils in some places and native dates in other.
Is it okay if i try to refactor this to use luxon only?
We definitely have to use the timezones here so I'll try to fix it either way, probably tomorrow or early next week.

MattIPv4 commented:
👍 Luxon is fine, we tend to use it anywhere where we need to deal with timezones.

bl4nkfce commented:
I've rebased the branch and pushed some changes.

- In general: updated some parts that used native JS Date objects in favor of Luxon DateTime objects.
- Updated the query that fetches events with the timezone info and tested some cases (end of month, end of day times).
- The calendar should correctly set the "current day" (and past days) highlight based on the selected timezone.
- Fixed Timezone Switch button text color in admin view (should now have better contrast with button bg, previously it inherited a very bright color)

@pjeweb @MattIPv4
If you could please test this thing when you have some time that would be great. I changed a bunch of things, so it's possible I broke something that worked and missed it during testing 😅 



https://github.com/user-attachments/assets/0a12508c-d63a-44d2-973a-11b8a89d5809

MattIPv4 commented on apps/website/src/components/calendar/Calendar.tsx:
```diff
@@ -383,20 +412,14 @@ const Calendar = ({
                         {/* Render the day of the week for mobile */}
                         <span className="md:hidden">
-                          {days[fullDate.getDay()]?.slice(0, 1)}
+                          {days[(i + startDay) % 7]?.charAt(0)}
```
You seem to be off by one here?

bl4nkfce replied:
I think it works now?  https://github.com/alveusgg/alveusgg/pull/744/commits/a15d770e22746bb3e48fa2ddbe960b747b039271 
The date used for selected month/year was not starting on the first day of the month, it was the current date instead hence incorrect weekday value.

MattIPv4 reviewed (dismissed):
Otherwise, LGTM

MattIPv4 reviewed (approved):
LGTM, tyvm.

pjeweb commented:
> @pjeweb did you want to review this before I merge it?

should be fine to merge. i'll take a look at it later though

bl4nkfce commented:
> Has anyone tried the extreme timezones (-12/+14)?

I just checked +14 and found two issues:
1. `.toLocaleTimeString()` returns this format for `KI-L` locale - we could just call `.toFormat('HH:mm')` to always display the time in a 24 hour format with padded mintues.
If not, we can leave it as is and rely on locale to take care of this formatting, but then it would be nice to do that for dates as well 🤔 
<img width="164" alt="image" src="https://github.com/user-attachments/assets/93de0de4-e0fb-4364-ac38-a948c666833d">

2. After switching from +14 to Alveus time, current day is not marked correctly (for some reason the "today" date is a month behind?)

Tbh I think my recent changes introduced more issues than solved. 😅 

> 
> One minor thing I just noticed. Would it be easier to understand if we'd call the local time "Your Time", since Alveus time is also local, just at Alveus...?

I like this idea, was also thinking about "My time (CEST)"

pjeweb reviewed (approved):
I think this is pretty solid now! If we wanna change the label for "my time", we can do that in a separate PR and get this merged first (finally).

pjeweb commented:
@bl4nkfce thank you again! are you in twitch chat too? or in one of the discords?

bl4nkfce commented:
yeah, I'm foszki on twitch and in Mayacord
